### ISSUE TITLE
Build is failing when some dependencies replaced with implementation or api instead of compile

### ISSUE THREAD

**Issue by lukavucko** [gradle/gradle#12204]
Hi,
I resolving deprecation for Gradle 7 and changing compile with implementation or api, but in some projects build failed when I changed compile with implementation/api.
Do you know why and what I should do?

Thank you in advance for your help.
Luka

**Comment by big-guy** [comment]
Without more details, it's hard to say where the problem could be.

The behavior is different when you're switching from `compile` to `api` or `implementation`.  If you make a dependency an `implementation` dependency, that means that downstream consumers won't see that dependency when they compile. 

So if you have two projects, A and B, and A depends on B and B depends on log4j, then when using `compile`, A and B would be compiled against log4j.  When using `implementation`, B would compile against log4j and A would not (because log4j is an "implementation detail" of B).  If A _needs_ log4j to compile, it should also directly depend on log4j.  Or if to _use_ interfaces from B, a consumer needs log4j to compile, B should add log4j as a `api` dependency.

In other words, you should use `implementation` if consumers of your library shouldn't care about the dependency to use your library.  You should use `api` if consumers need to use types from the dependency to use your library. 

What you may be seeing is that there's a dependency that one project has that the others have also used without declaring a dependency on it.  If you've made that dependency an implementation dependency, the other projects will no longer get that dependency transitively.

**Comment by lukavucko** [comment]
Thank you for your explanation, but I thought that compile is equivalent with api.
In some projects when I switched first to implementation, build failed, then I switched with api and build failed again, this is strange and I asking you maybe you know why?

**Comment by ljacomet** [comment]
`api` is not equivalent to `compile` however switching all dependencies to from `compile` to `api`, while not recommended, should indeed not result in missing dependencies.

But since you provide limited information on what the failure is, or the specifics of your build, it is hard to be anything but generic in the answer.

**Comment by lukavucko** [comment]
For example project A use implementation dependency for something.
 implementation(':something').
This something also should be used for implementation in project B, C and others?

**Comment by ljacomet** [comment]
No, `implementation` dependencies are only transitive for the `runtime` classpath of consumers, for `compile` classpath, they are hidden, as _implementation_ details. That is the concept of the feature.

**Comment by lukavucko** [comment]
Thank you, is it possible to detect for example error as org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':projectName:compileJava' where is the causes, implementation or api and on which places?
This is only when I switched implementation or api instead of compile, but for example if you switched one dependency with implementation and then executed the build and you know if it's failed or successfull, but if you have many projects and for example switched everything with implementation, not one by on, how you can know wheere is the problem if you have error like above?
